src/objects/Arrayable.ts

```diff
@@ -147,15 +147,17 @@ class Arrayable<T> extends Array<T> {
         return this.constructor.from(clonedeep([...this])) as Arrayable<T>
     }
 
-    groupBy(key: string | Function) {
-        const grouped = this.reduce<{ [key: string]: Arrayable<T> }>((result, item) => {
+    groupBy<K>(key: string | Function) {
+        return this.reduce<Mappable<K, Arrayable<T>>>((result, item) => {
             const group = typeof key === "function" ? key(item) : item[key]
-            result[group] = result[group] || new this.constructor()
-            result[group].push(item)
+            if (result.has(group)) {
+                const newItem = new this.constructor(...result.get(group), item)
```

Comment:
Since we create the array in here, we can safely push into it

```suggestion
                result.get(group).push(item)
```

Then the next line is not needed as well.